### PR TITLE
bpo-32388: Remove cross-version binary compatibility requirement in tp_flags

### PR DESCRIPTION
**Pull request by pitrou**
Starting with 3.8, it is now allowed to add new fields at the end of the `PyTypeObject` struct in feature releases without having to allocate a dedicated compatibility flag in `tp_flags`. This will reduce the risk of running out of bits in the 32-bit `tp_flags` value.

https://bugs.python.org/issue32388

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -675,8 +672,14 @@ given type object has a specified feature.
 /* NOTE: The following flags reuse lower bits (removed as part of the
  * Python 3.0 transition). */
 
+/* The following flags are kept for compatibility, but all types are
+ * supposed to support the given functionality.  Starting with 3.7,
```
Comment (scoder) [inline Include/object.h]: This explicitly mentions the CPython release of the change and needs adaptation when this gets merged.

Comment (JimJJewett) [reply Include/object.h]: I think it should also be mentioned that Py_TPFLAGS_HAVE_FINALIZE is being rolled into the always-supported default, if that effects how to write objects.  If it does not affect anything extension writers do, then that (and the reason it doesn't) should be mentioned.

File: Objects/object.c [hunk 2]
```diff
@@ -276,8 +276,7 @@ PyObject_CallFinalizer(PyObject *self)
 
     /* The former could happen on heaptypes created from the C API, e.g.
        PyType_FromSpec(). */
```
Comment (JimJJewett) [inline Objects/object.c]: I think this comment (including the existence of a "former" depends on removed code, and should therefore itself be removed.

File: Include/object.h (hunk 1 above)
Comment (JimJJewett) [reply Include/object.h]: Conversely, if it does affect anything for extension writers -- even just simplifying things -- that should be added to the porting section of what's new.

**Comment by jdemeyer** [comment]
You write "Starting with 3.7, binary compatibility of C extensions accross feature releases of Python is not supported anymore" but has such binary compatibility ever been supported? This PR seems to imply that it was, but PEP 384 explicitly says that the layout of `PyTypeObject` is not part of the stable ABI.

PS: typo `accross` -> `across`

**Comment by pitrou** [comment]
I updated this PR against master and addressed the review comments.

**Comment by pitrou** [comment]
@zooba Is there a way to retry an Azure job which failed because it couldn't "provision the agent"?

File: Doc/whatsnew/3.8.rst [hunk 3]
```diff
@@ -1019,6 +1019,13 @@ Changes in the C API
 
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The interpreter does not pretend to support binary compatibility of
+  extension types accross feature releases, anymore.  A :c:type:`PyTypeObject`
+  exported by a third-party extension module is supposed to have all the
+  slots expected in the current Python version, including
+  :c:member:`~PyTypeObject.tp_finalize` (:const:`Py_TPFLAGS_HAVE_FINALIZE`
+  is not checked anymore before reading :c:member:`~PyTypeObject.tp_finalize`).
```
Comment (tirkarthi) [inline Doc/whatsnew/3.8.rst:1269]: Could probably link to bpo with :issue:`32388` since it's in What's New page.

File: Include/object.h [hunk 4]
```diff
@@ -332,8 +329,14 @@ given type object has a specified feature.
 /* NOTE: The following flags reuse lower bits (removed as part of the
  * Python 3.0 transition). */
 
+/* The following flags are kept for compatibility, but all types are
```
Comment (jdemeyer) [inline Include/object.h]: `flags are` -> `flag is`

**Comment by jdemeyer** [comment]
It would be good to get this done before the first 3.8 beta.

@scoder: is there any chance that you could finish the "core review"?

**Comment by jdemeyer** [comment]
> Starting with 3.8, it is now allowed to add new fields at the end of the PyTypeObject struct (and even in the middle, if we are so inclined)

Not "in the middle" as that would break the typical code of the form
```C
PyTypeObject PyCFunction_Type = {
    PyVarObject_HEAD_INIT(&PyType_Type, 0)
    "builtin_function_or_method",
    sizeof(PyCFunctionObject),
    0,
    (destructor)meth_dealloc,                   /* tp_dealloc */
    0,                                          /* tp_print */
    0,                                          /* tp_getattr */
    0,                                          /* tp_setattr */
    0,                                          /* tp_reserved */
    (reprfunc)meth_repr,                        /* tp_repr */
    0,                                          /* tp_as_number */
    0,                                          /* tp_as_sequence */
    0,                                          /* tp_as_mapping */
    (hashfunc)meth_hash,                        /* tp_hash */
    PyCFunction_Call,                           /* tp_call */
    0,                                          /* tp_str */
    PyObject_GenericGetAttr,                    /* tp_getattro */
    0,                                          /* tp_setattro */
    0,                                          /* tp_as_buffer */
    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,/* tp_flags */
    0,                                          /* tp_doc */
    (traverseproc)meth_traverse,                /* tp_traverse */
    0,                                          /* tp_clear */
    meth_richcompare,                           /* tp_richcompare */
    offsetof(PyCFunctionObject, m_weakreflist), /* tp_weaklistoffset */
    0,                                          /* tp_iter */
    0,                                          /* tp_iternext */
    meth_methods,                               /* tp_methods */
    meth_members,                               /* tp_members */
    meth_getsets,                               /* tp_getset */
    0,                                          /* tp_base */
    0,                                          /* tp_dict */
};
```
With C99, there are better ways to write the above, but we should still support that.

**Comment by jdemeyer** [comment]
One thing that keeps bugging me when I see this diff: Do we really have to change the value of `Py_TPFLAGS_HAVE_FINALIZE`? I don't see any immediate problem with changing the value, but it would be more prudent to keep it as `1 << 0`.

**Comment by vstinner** [comment]
cc @encukou

**Comment by encukou** [comment]
Can we keep the bit for backwards compatibility?
We're not running out of bits any time soon, especially since we'll not be adding one for each new feature. (Even though I want to [add two of them now](https://www.python.org/dev/peps/pep-0590/).)

Pedantically, something like `MY_USUAL_FLAGS & ~Py_TPFLAGS_HAVE_FINALIZE` would be correct in 3.7, so it should work in 3.8 as well.

**Comment by pitrou** [comment]
> Pedantically, something like MY_USUAL_FLAGS & ~Py_TPFLAGS_HAVE_FINALIZE would be correct in 3.7, so it should work in 3.8 as well.

What value do you suggest `Py_TPFLAGS_HAVE_FINALIZE` should have then?

**Comment by jdemeyer** [comment]
> What value do you suggest `Py_TPFLAGS_HAVE_FINALIZE` should have then?

The same value that it always had.

**Comment by pitrou** [comment]
Ok. Unless someone opposes, I'll merge this once CI is green.

**Comment by encukou** [comment]
Thank you!

**Comment by scoder** [comment]
Nice, happy to see that this was merged.

**Pull request merged**